feat(table): Basic Transaction and AddFiles

zeroshade opened the pull request:
And here we go! Snapshot producers, basic Transaction object, and a basic implementation of adding files to a table!

Complete with unit tests and a general TableWritingTestSuite.

zeroshade commented:
We're in the home stretch @Fokko @kevinjqliu!! Thanks so much for the quick reviews and feedback on all of these.

kevinjqliu commented on table/table_test.go:
```diff
@@ -128,3 +138,235 @@ func (t *TableTestSuite) TestSnapshotByName() {
+}
+
+func TestTableWriting(t *testing.T) {
+	suite.Run(t, &TableWritingTestSuite{formatVersion: 1})
```
also run for V2 tables?

zeroshade replied:
bah, made it into a subsequent commit that wasn't part of this one. -_- I'll update!

Fokko commented on table/metadata.go:
```diff
@@ -214,6 +234,37 @@ func (b *MetadataBuilder) CurrentSchema() *iceberg.Schema {
+
+func (b *MetadataBuilder) newSnapshotID() int64 {
+	snapshotID := generateSnapshotID()
+	for slices.ContainsFunc(b.snapshotList, func(s Snapshot) bool { return s.SnapshotID == snapshotID }) {
+		snapshotID = generateSnapshotID()
+	}
```
This is a pretty neat syntax!

Fokko commented on table/arrow_utils.go:
```diff
@@ -908,6 +915,260 @@ func must[T any](v T, err error) T {
+		return "FIXED_LEN_BYTE_ARRAY"
+	case iceberg.BinaryType:
+		return "BYTE_ARRAY"
+	case iceberg.DecimalType:
+		return "FIXED_LEN_BYTE_ARRAY"
```
For the decimal, it depends on the length

![image](https://github.com/user-attachments/assets/51c3a4b9-8966-4ad4-bef0-a63b6bb3f5e4)

zeroshade replied:
When I implement the writing to Parquet, the go parquet library already handles mapping Decimal values to int32 or int64 when converting the Arrow -> Parquet schemas

Fokko replied:
Yes, that's a bug in PyIceberg too :) Thanks for catching that!

Fokko commented on table/transaction.go:
```diff
@@ -0,0 +1,346 @@
+	}
+
+	if checkDuplicates {
+		// TODO: implement
```
By exposing this, an user might assume that this is already implemented. Maybe we should raise here?

zeroshade replied:
In a subsequent PR I implement this. But you're right, I'll have it return a not implemented error here for now.

Fokko commented on table/arrow_utils.go:
```diff
@@ -1250,5 +1566,56 @@ func dataFileStatsFromParquetMetadata(pqmeta *metadata.FileMetaData, statsCols m
+	}
+}
+
+func parquetFilesToDataFiles(fileIO iceio.IO, meta *MetadataBuilder, paths iter.Seq[string]) iter.Seq2[iceberg.DataFile, error] {
```
I'm reluctant to add this with this PR. Do we handle the name-mapping properly?

zeroshade replied:
yup, name mapping is handled properly (as far as I can tell via testing) though over the course of writing tests for the subsequent changes, I did find a couple edge cases which I fixed in those commits. They'll be included in the next PR after this one.

Fokko commented on table/properties.go:
```diff
@@ -25,10 +25,24 @@ const (
 
-	DefaultNameMappingKey = "schema.name-mapping.default"
-
+	DefaultNameMappingKey       = "schema.name-mapping.default"
```
nit: I think it was better before :)

zeroshade replied:
keeping the empty space between them instead of having it part of the same group? sure I can put it back :)

Fokko commented on table/snapshot_producers.go:
```diff
@@ -0,0 +1,582 @@
+			unmergedDataManifests = append(unmergedDataManifests, m)
+		} else if m.ManifestContent() == iceberg.ManifestContentDeletes {
+			unmergedDeleteManifests = append(unmergedDeleteManifests, m)
+		}
```
What's the behavior here when it doesn't match? should we throw?

zeroshade replied:
I was following the pyiceberg example [here](https://github.com/apache/iceberg-python/blob/a275ce5941dc85b1721562ea4d2e316e2b36b28e/pyiceberg/table/update/snapshot.py#L144) looks like manifests which don't match are just ignored.

Fokko commented on table/transaction.go:
```diff
@@ -0,0 +1,346 @@
+		// TODO: implement
+	}
+
+	if t.meta.NameMapping() == nil {
```
If new columns are added, should we also extend the name-mapping? Or do we throw then?

zeroshade replied:
`AddFiles` doesn't allow changing the schema by adding columns, I haven't added and `UpdateSchema` method yet for adding new columns etc.

Fokko replied:
I see, then this is okay, once we start changing schema's it is important to update this:

ref: https://github.com/apache/iceberg-python/blob/a275ce5941dc85b1721562ea4d2e316e2b36b28e/pyiceberg/table/update/schema.py#L546-L550

Fokko commented on table/table_test.go:
```diff
@@ -128,3 +138,236 @@ func (t *TableTestSuite) TestSnapshotByName() {
+	files := make([]string, 0)
+	for i := range 5 {
+		filePath := fmt.Sprintf("%s/unpartitioned/test-%d.parquet", t.location, i)
+		t.writeParquet(tbl.FS().(iceio.WriteFileIO), filePath, t.arrTbl)
```
I'm confused here; shouldn't the `t` produce Parquet files with IDs? The `AddFiles` depends on `parquetFilesToDataFiles` where we check if it has IDs.

Fokko replied:
Ah okay, I thought this was part of Iceberg. The writers provided by Iceberg always should produce IDs

zeroshade replied:
Anything in a file that ends with "_test.go" is only for tests :)

When I implement the writers it will definitely write IDs

Fokko commented on table/transaction.go:
```diff
@@ -0,0 +1,346 @@
+	return nil
+}
+
+func (t *Transaction) Append(rdr array.RecordReader, snapshotProps iceberg.Properties) error {
```
This doesn't anything, right?

zeroshade replied:
correct, currently it doesn't do anything yet. I'll change it to return a not implemented error

Fokko reviewed (changes requested):
Thanks Matt for working on this, some great progress here!

There is quite a bit code here, but if I understand it correctly. We first write Parquet files, and then add them to the table using `AddFiles`. I think this is wrong, since `AddFiles` is meant as a migration feature. This method will fetch the footer, check the schema etc. Also, the writers that come with Iceberg-go, should always produce the Field-IDs.

Instead, when we write Parquet, we directly want to convert that into a `DataFile`, and then we can add them to the table using `add_file`: https://github.com/apache/iceberg-python/blob/8b2d3ec99a7d177423bd110e095bacf0fba87adb/pyiceberg/table/snapshots.py#L121

I think it would be good to split this into two PRs:

- First one PR that adds the `Transaction`, including the fast- and merge-append operation etc.
- A separate PR for the `AddFiles` since that comes with a lot of caveats.

zeroshade commented:
> There is quite a bit code here, but if I understand it correctly. We first write Parquet files, and then add them to the table using `AddFiles`. I think this is wrong, since `AddFiles` is meant as a migration feature. This method will fetch the footer, check the schema etc. Also, the writers that come with Iceberg-go, should always produce the Field-IDs.

This *only* implements the migration feature of `AddFiles`. This does not yet implement any writing of parquet files by `iceberg-go` outside of the unit tests writing parquet files without field ids for testing the `AddFiles` function.

> Instead, when we write Parquet, we directly want to convert that into a DataFile, and then we can add them to the table using add_file

Yup, i plan on doing that as future work. It was an easier code path to put together the `AddFiles` migration functions than to implement all of the work for splitting up records and tables for determining when and how to write out the parquet files. 

When I changed the `Append` to just return a NotImplemented error a whole bunch of code for the manifest merging was now unused, so I've removed that from this PR to simplify it more. Is that sufficient to reduce this to make it easier to review? Otherwise I can look into splitting it more along what you've suggested.

zeroshade commented:
@Fokko The difficulty with splitting out the AddFiles portion to a separate PR is that all of the table writing unit tests that I added rely on AddFiles being implemented for actually testing the whole way through. I can't figure out a good way to easily test the internal manifest generation and handling without doing a full end-to-end test by actually adding files to the table.

What do you think?

zeroshade commented:
The easiest part to excise as a separate PR is likely the stats aggregation implementation pieces. I can remove that from this PR (most of the tests should still work fine for adding files, just without the bounds existing in the result) and then I can create a PR for just the stats aggregations. Would that shrink this enough?

zeroshade commented on table/arrow_utils_internal_test.go:
```diff
@@ -408,3 +478,158 @@ func TestStatsTypes(t *testing.T) {
+                    {"id": 10, "name": "strings", "required": false, "type": "string"},
+                    {"id": 11, "name": "uuids", "required": false, "type": "uuid"},
+                    {"id": 12, "name": "binaries", "required": false, "type": "binary"}
+                ]
```
test cases added for 3 decimal values with precision of 8, 16, and 24.

zeroshade commented:
@Fokko I've added the test cases and handling of decimals. Do you still need me to split this PR into smaller chunks or anything else? Or are we good here and ready for the next one? :smile:

kevinjqliu reviewed (approved):
LGTM, this is a quick way to get parquet files into an Iceberg table. 

The `AddFiles` API is properly tested in `table_test.go` and the `DataFile`/`statsAggregator` is properly tested in `arrow_utils_internal_test.go`

Most of the tests for the table transaction and commit logic is done via `AddFiles`, so I'm not sure splitting the PR for just the Transaction piece is useful. I'll defer to @Fokko to chime in on that part

zeroshade commented on table/transaction.go:
```diff
@@ -0,0 +1,340 @@
+	return compat
+}
+
+func checkArrowSchemaCompat(requested *iceberg.Schema, provided *arrow.Schema, downcastNanoToMicro bool) error {
```
shifted it to arrow_utils.go

zeroshade commented on table/transaction.go:
```diff
@@ -0,0 +1,340 @@
+	return newFastAppendFilesProducer(OpAppend, s.txn, s.io, nil, s.snapshotProps)
+}
+
+// func (s snapshotUpdate) mergeAppend() *snapshotProducer {
```
removed

zeroshade commented on table/transaction.go:
```diff
@@ -0,0 +1,340 @@
+		}
+	}
+
+	prevUpdates, prevLastUpdated := len(t.meta.updates), t.meta.lastUpdatedMS
+	for _, u := range updates {
+		if err := u.Apply(t.meta); err != nil {
+			return err
+		}
+	}
+
+	if prevUpdates < len(t.meta.updates) {
```
calling `u.Apply(t.meta)` will potentially add to `t.meta.updates` if the requested updates are valid and not-noops.

this is basically equivalent to the `HasChanges` check in pyiceberg. We're looking at the current list of updates, then potentially adding new updates. So if `Apply` results in creating changes, `len(t.meta.updates)` will be larger than `prevUpdates`

zeroshade replied:
I've added a comment to explain this in the code

kevinjqliu commented on table/arrow_utils.go:
```diff
@@ -1250,5 +1656,56 @@ func dataFileStatsFromParquetMetadata(pqmeta *metadata.FileMetaData, statsCols m
+				must(computeStatsPlan(meta.CurrentSchema(), meta.props)),
+				must(parquetPathToIDMapping(meta.CurrentSchema())))
+
+			df, err := statistics.toDataFile(meta.CurrentSpec(), filePath, iceberg.ParquetFile, rdr.MetaData().GetSourceFileSize())
```
`toDataFile` is using a builder so this is not a problem

Fokko commented on table/table.go:
```diff
@@ -72,6 +76,64 @@ func (t Table) LocationProvider() (LocationProvider, error) {
+	return &Transaction{
+		tbl:  &t,
+		meta: meta,
+		reqs: []Requirement{},
```
Why wouldn't you keep the updates and requirements next to each other in the Transaction?

Later on, when we implement retries, the metadata might refresh if a table has been updated. Then we can detect if we can re-apply the changes. Therefore I think it is better to keep this at the Transaction level.

Fokko reviewed (approved):
One suggestion, but apart from that this looks good. Thanks @zeroshade and thanks for the review @kevinjqliu
